Index: AMEGIC++/Main/Process_Group.C
===================================================================
--- AMEGIC++/Main/Process_Group.C	(revision 28115)
+++ AMEGIC++/Main/Process_Group.C	(working copy)
@@ -52,7 +52,7 @@
   if (typechk>1) THROW(fatal_error,"NLO_QCD_Parts 'RS', and 'BVI' must be assigned separately!");
 
   nlo_type::code nloqcd=pi.m_fi.m_nloqcdtype;
-  if (nloqcd&nlo_type::real && nloqcd&nlo_type::rsub) {
+  if ((nloqcd&nlo_type::real) || (nloqcd&nlo_type::rsub)) {
     Single_Real_Correction *src = new Single_Real_Correction();
     src->SetNoTree(pi.m_rsmegenerator.length() &&
 		   pi.m_rsmegenerator!="Amegic");
Index: PHASIC++/Process/MCatNLO_Process.C
===================================================================
--- PHASIC++/Process/MCatNLO_Process.C	(revision 28115)
+++ PHASIC++/Process/MCatNLO_Process.C	(working copy)
@@ -82,7 +82,7 @@
   p_rproc=InitProcess(spi,nlo_type::lo,true);
   spi.m_megenerator=pi.m_megenerator;
   p_bviproc=InitProcess(spi,nlo_type::born|nlo_type::loop|nlo_type::vsub,false);
-  p_ddproc=InitProcess(spi,nlo_type::real|nlo_type::rsub,1);
+  p_ddproc=InitProcess(spi,nlo_type::rsub,1);
   spi.m_integrator=spi.m_rsintegrator;
   spi.m_megenerator=spi.m_rsmegenerator;
   spi.m_itmin=spi.m_rsitmin;
