Index: AMEGIC++/DipoleSubtraction/Single_LOProcess.C
===================================================================
--- AMEGIC++/DipoleSubtraction/Single_LOProcess.C	(revision 28115)
+++ AMEGIC++/DipoleSubtraction/Single_LOProcess.C	(working copy)
@@ -38,11 +38,7 @@
   m_nin=pi.m_ii.NExternal();
   m_nout=pi.m_fi.NExternal();
 
-  Data_Reader reader(" ",";","!","=");
-  reader.AddComment("#");
-  reader.SetInputPath(rpa->GetPath());
-  reader.SetInputFile(rpa->gen.Variable("ME_DATA_FILE"));
-  int ord=reader.GetValue<int>("AMEGIC_SORT_LOPROCESS",1);
+  int ord=ToType<int>(rpa->gen.Variable("AMEGIC_SORT_LOPROCESS"));
   static bool print(false);
   if (!print && !ord) {
     print=true;
@@ -200,14 +196,11 @@
   p_hel    = new Helicity(m_nin,m_nout,&m_flavs.front(),p_pl);
 
   bool directload = true;
-  int libchk=0; 
-  Data_Reader reader(" ",";","!","=");
-  if (reader.ReadFromFile(libchk,"ME_LIBCHECK")) {
+  int libchk=ToType<int>(rpa->gen.Variable("ME_LIBCHECK")); 
     if (libchk==1) {
       msg_Info()<<"Enforce full library check. This may take some time"<<std::endl;
       directload = false;
     }
-  }  
   if (directload) directload = FoundMappingFile(m_libname,m_pslibname);
   if (directload) {
     string hstr=rpa->gen.Variable("SHERPA_CPP_PATH")+"/Process/Amegic/"+m_ptypename+"/"+m_libname;
@@ -218,7 +211,7 @@
   p_BS->Setk0(s_gauge);
   p_shand  = new String_Handler(m_gen_str,p_BS,model->p_model->GetCouplings());
   int ntchanmin(m_ntchanmin);
-  bool cvp(reader.GetValue<int>("AMEGIC_CUT_MASSIVE_VECTOR_PROPAGATORS",1));
+  bool cvp(ToType<int>(rpa->gen.Variable("AMEGIC_CUT_MASSIVE_VECTOR_PROPAGATORS")));
   p_ampl   = new Amplitude_Handler(m_nin+m_nout,&m_flavs.front(),p_b,p_pinfo,model,top,m_maxcpl,m_mincpl,ntchanmin,
                                    &m_cpls,p_BS,p_shand,m_print_graphs,!directload,cvp);
   m_ntchanmin=ntchanmin;
Index: AMEGIC++/Main/Amegic.C
===================================================================
--- AMEGIC++/Main/Amegic.C	(revision 28115)
+++ AMEGIC++/Main/Amegic.C	(working copy)
@@ -117,6 +117,12 @@
   p_int->SetBeam(beamhandler);
   p_int->SetISR(isrhandler);
   SetPSMasses(&read);
+  int sort=read.GetValue<int>("AMEGIC_SORT_LOPROCESS",1);
+  rpa->gen.SetVariable("AMEGIC_SORT_LOPROCESS",ToString(sort));
+  int libcheck=read.GetValue<int>("ME_LIBCHECK",0);
+  rpa->gen.SetVariable("ME_LIBCHECK",ToString(libcheck));
+  int cvp=read.GetValue<int>("AMEGIC_CUT_MASSIVE_VECTOR_PROPAGATORS",1);
+  rpa->gen.SetVariable("AMEGIC_CUT_MASSIVE_VECTOR_PROPAGATORS",ToString(cvp));
   double alpha=read.GetValue<double>("AMEGIC_TCHANNEL_ALPHA",0.9);
   rpa->gen.SetVariable("AMEGIC_TCHANNEL_ALPHA",ToString(alpha));
   double salpha=read.GetValue<double>("AMEGIC_SCHANNEL_ALPHA",0.75);
